File: 1497-check-if-array-pairs-are-divisible-by-k/1497-check-if-array-pairs-are-divisible-by-k.py
from collections import Counter
from typing import List

class Solution:
    def canArrange(self, arr: List[int], k: int) -> bool:
        freq = Counter((x % k + k) % k for x in arr)
        
        for remainder in freq:
            if remainder == 0:
                if freq[remainder] % 2 != 0:
                    return False
            elif remainder * 2 == k:
                if freq[remainder] % 2 != 0:
                    return False
            else:
                if freq[remainder] != freq[k - remainder]:
                    return False
                    
        return True

File: 1497-check-if-array-pairs-are-divisible-by-k/1497-check-if-array-pairs-are-divisible-by-k.java
class Solution {
    public boolean canArrange(int[] arr, int k) {
        int[] freq = new int[k];

        for (int x : arr) {
            int rem = ((x % k) + k) % k;
            freq[rem]++;
        }

        if (freq[0] % 2 != 0) {
            return false;
        }

        for (int i = 1; i <= k / 2; i++) {
            if (i == k - i) {
                if (freq[i] % 2 != 0) {
                    return false;
                }
            } else {
                if (freq[i] != freq[k - i]) {
                    return false;
                }
            }
        }

        return true;
    }
}